tc_build/llvm.py

```diff
@@ -0,0 +1,545 @@
+        self.cmake_defines['LLVM_INCLUDE_TESTS'] = 'OFF'
+
+        super().configure()
+
+
+class LLVMInstrumentedBuilder(LLVMBuilder):
+
+    def __init__(self):
+        super().__init__()
+
+        self.cmake_defines['LLVM_BUILD_INSTRUMENTED'] = 'IR'
+        self.cmake_defines['LLVM_BUILD_RUNTIME'] = 'OFF'
+        self.cmake_defines['LLVM_LINK_LLVM_DYLIB'] = 'ON'
+
+    def configure(self):
+        # The following defines are needed to avoid thousands of warnings
+        # along the lines of:
+        # "Unable to track new values: Running out of static counters."
+        # They require LLVM_LINK_DYLIB to be enabled, which is done above.
+        cmake_options = Path(self.folders.source, 'llvm/cmake/modules/HandleLLVMOptions.cmake')
+        cmake_text = cmake_options.read_text(encoding='utf-8')
+        if 'LLVM_VP_COUNTERS_PER_SITE' in cmake_text:
+            self.cmake_defines['LLVM_VP_COUNTERS_PER_SITE'] = '6'
+        else:
+            cflags = []
+            cxxflags = []
+
+            if 'CMAKE_C_FLAGS' in self.cmake_defines:
+                cflags += self.cmake_defines['CMAKE_C_FLAGS'].split(' ')
+            if 'CMAKE_CXX_FLAGS' in self.cmake_defines:
+                cxxflags += self.cmake_defines['CMAKE_CXX_FLAGS'].split(' ')
+
+            vp_counters = [
+                '-Xclang',
+                '-mllvm',
+                '-Xclang',
+                '-vp-counters-per-site=6',
+            ]
+            cflags += vp_counters
+            cxxflags += vp_counters
+
+            self.cmake_defines['CMAKE_C_FLAGS'] = ' '.join(cflags)
+            self.cmake_defines['CMAKE_CXX_FLAGS'] = ' '.join(cxxflags)
+
+        super().configure()
+
+    def generate_profdata(self):
+        if not (profiles := list(self.folders.build.joinpath('profiles').glob('*.profraw'))):
+            raise RuntimeError('No profiles generated?')
+
+        llvm_prof_data_cmd = [
+            self.tools.llvm_profdata,
+            'merge',
+            f"-output={Path(self.folders.build, 'profdata.prof')}",
+            *profiles,
+        ]
+        subprocess.run(llvm_prof_data_cmd, check=True)
+
+
+class LLVMSlimInstrumentedBuilder(LLVMInstrumentedBuilder, LLVMSlimBuilder):
+    # No methods to override, this class inherits everyting from these super classes
+    pass
+
+
+class LLVMSourceManager:
+
+    def __init__(self, repo):
+        self.repo = repo
+
+    def default_projects(self):
+        return ['clang', 'compiler-rt', 'lld', 'polly']
+
+    def default_targets(self):
+        return ['AArch64', 'ARM', 'BPF', 'Hexagon', 'Mips', 'PowerPC', 'RISCV', 'SystemZ', 'X86']
+
+    def download(self, ref, shallow=False):
+        if self.repo.exists():
+            return
+
+        utils.print_header('Downloading LLVM')
+
+        git_clone = ['git', 'clone']
+        if shallow:
+            git_clone.append('--depth=1')
+            if ref != 'main':
+                git_clone.append('--no-single-branch')
```

Comment:
Do you mean for this to be conditional on `shallow`?

Reply:
This is in an `if shallow` block already.